lib/http/server.js

```diff
@@ -0,0 +1,134 @@
+'use strict';
+
+const {
+  Promise,
+  RegExpPrototypeExec,
+  StringPrototypeEndsWith,
+  StringPrototypeIndexOf,
+  StringPrototypeLastIndexOf,
+  StringPrototypeReplace,
+  StringPrototypeSlice,
+  StringPrototypeStartsWith,
+  StringPrototypeToLocaleLowerCase,
+} = primordials;
+
+const { cwd, nextTick } = process;
+const console = require('internal/console/global');
+const { createReadStream } = require('fs');
+const { Server } = require('_http_server');
+const { URL, isURLInstance, pathToFileURL } = require('internal/url');
+const { sep } = require('path');
+const { emitExperimentalWarning } = require('internal/util');
+
+emitExperimentalWarning('http/server');
+
+if (module.isPreloading) {
+  const { parseArgs } = require('internal/util/parse_args/parse_args');
+  const { values: { port, host } } = parseArgs({ options: {
+    port: {
+      type: 'string',
+      short: 'p',
+    },
+    host: {
+      type: 'string',
+      short: 'h',
+    },
+  } });
+  nextTick(createSimpleServer, process.argv[1], port, host);
+  process.env.NODE_REPL_EXTERNAL_MODULE = 'node:http/server';
+  process.argv[1] = 'node:http/server';
+}
+
+function createSimpleServer(directory = cwd(), port = 0, host = 'localhost') {
+  const CWD_URL = isURLInstance(directory) ? directory : pathToFileURL(directory + sep);
+
+  const server = new Server(async (req, res) => {
+    const url = new URL(
+      req.url === '/' || StringPrototypeStartsWith(req.url, '/?') ?
+        './index.html' :
+        StringPrototypeReplace(new URL(req.url, 'root://').toString(), 'root://', '.'),
+      CWD_URL
+    );
+
+    const mime = {
+      '__proto__': null,
+      '.html': 'text/html; charset=UTF-8',
+      '.js': 'text/javascript; charset=UTF-8',
+      '.css': 'text/css; charset=UTF-8',
+      '.avif': 'image/avif',
+      '.svg': 'image/svg+xml',
+      '.jpg': 'image/jpeg',
+      '.jpeg': 'image/jpeg',
+      '.png': 'image/png',
+      '.webp': 'image/webp',
+      '.woff2': 'font/woff2',
+    };
+    const ext = StringPrototypeToLocaleLowerCase(
+      StringPrototypeSlice(url.pathname, StringPrototypeLastIndexOf(url.pathname, '.'))
+    );
+    if (ext in mime) res.setHeader('Content-Type', mime[ext]);
+
+    try {
+      try {
+        const stream = createReadStream(url, { emitClose: false });
+        await new Promise((resolve, reject) => {
+          stream.on('error', reject);
+          stream.on('close', resolve);
```

Comment:
Nit: keep in mind that `'close'` is always emitted after `'error'`. If the `'error'` event is emitted and the `reject` is called, then the `'close'` event should not call `resolve`.

Reply:
Not a significant problem at all... just a non-blocking nit. Feel free to ignore.